android/src/main/java/com/polidea/reactnativeble/errors/ErrorConverter.java

```diff
@@ -1,58 +1,212 @@
+            BleError bleError = new BleError(BleErrorCode.DescriptorReadFailed, message, code);
+            bleError.deviceID = deviceID;
+            bleError.serviceUUID = serviceUUID;
+            bleError.characteristicUUID = characteristicUUID;
+            bleError.descriptorUUID = descriptorUUID;
+            return bleError;
+        } else if (BleGattOperationType.DESCRIPTOR_WRITE == operationType) {
+            BleError bleError = new BleError(BleErrorCode.DescriptorWriteFailed, message, code);
+            bleError.deviceID = deviceID;
+            bleError.serviceUUID = serviceUUID;
+            bleError.characteristicUUID = characteristicUUID;
+            bleError.descriptorUUID = descriptorUUID;
+        } else if (BleGattOperationType.READ_RSSI == operationType) {
+            BleError bleError = new BleError(BleErrorCode.DeviceRSSIReadFailed, message, code);
+            bleError.deviceID = deviceID;
+            return bleError;
+        } else if (BleGattOperationType.ON_MTU_CHANGED == operationType) {
+            BleError bleError = new BleError(BleErrorCode.DeviceMTUChangeFailed, message, code);
+            bleError.deviceID = deviceID;
+            return bleError;
+        } else if (BleGattOperationType.CONNECTION_PRIORITY_CHANGE == operationType) {
+            // TODO: Handle?
         }
-        return new Error("Unknown error: " + throwable.toString(), 0);
+
+        return new BleError(BleErrorCode.UnknownError, message, code);
     }
 
-    private Error toError(BleScanException bleScanException) {
+    private BleError toError(BleScanException bleScanException) {
         final int reason = bleScanException.getReason();
         switch (reason) {
             case BleScanException.BLUETOOTH_CANNOT_START:
-                return new Error("Bluetooth cannot start", 103);
+                return new BleError(BleErrorCode.ScanStartFailed, bleScanException.getMessage(), null);
             case BleScanException.BLUETOOTH_DISABLED:
-                return new Error("Bluetooth is powered off", 102);
+                return new BleError(BleErrorCode.BluetoothPoweredOff, bleScanException.getMessage(), null);
             case BleScanException.BLUETOOTH_NOT_AVAILABLE:
-                return new Error("Bluetooth is not supported", 100);
+                return new BleError(BleErrorCode.BluetoothUnsupported, bleScanException.getMessage(), null);
             case BleScanException.LOCATION_PERMISSION_MISSING:
-                return new Error("Bluetooth location permission is missing", 110);
+                return new BleError(BleErrorCode.BluetoothUnauthorized, bleScanException.getMessage(), null);
             case BleScanException.LOCATION_SERVICES_DISABLED:
-                return new Error("Bluetooth location services are disabled", 111);
+                return new BleError(BleErrorCode.BluetoothUnauthorized, bleScanException.getMessage(), null);
```

Comment:
This should be probably a different code? It may be not convenient to parse the message?

Reply:
LocationServices are implementation specific and ErrorCode is designed to be platform agnostic. In ideal case this should set androidErrorCode, but this won't happen as it is unrelated to Gatt. I'm not sure how to solve it.

Reply:
Probably the best idea is just to add a new custom code under some (to be decided) label. It should definitely be distinguishable from the Missing Permission case

Reply:
Agree!